File: src/firmware/uplink.ts
/**
 * HTTP uplink to HomeHub.
 *
 * Posts each reading to `${HOMEHUB_URL}/internal/sensors/flow/readings` with
 * `X-Internal-Key`. On any failure, buffers the reading in a ring (~5 min
 * capacity) and retries on the next push, oldest first. Each buffered
 * reading carries its enqueue time so the server can reconstruct the
 * actual recordedAt via `tOffsetMs`.
 *
 * Server response carries the desired poll cadence — the orchestrator
 * applies it on the next loop iteration. No HTTP server needed on-device.
 */

import { RingBuffer } from './ring-buffer';

export interface UplinkConfig {
  homehubUrl:      string;          // e.g. "http://192.168.1.10:3000"
  internalApiKey:  string;
  sensorId:        string;
  bufferCapacity?: number;
  /** Absolute request timeout. */
  timeoutMs?:      number;
}

export interface ReadingPayload {
  rateM3h:           number;
  totalM3?:          number | null;
  signalQuality?:    number | null;
  rssi?:             number | null;
  uptimeSec?:        number | null;
  firmwareVersion?:  string | null;
  mac?:              string | null;
  bootReason?:       'power' | 'software' | 'watchdog' | 'unknown' | null;
}

export interface PollConfig {
  pollIntervalMs:        number;
  flowingPollIntervalMs: number;
  idlePollIntervalMs:    number;
  configVersion:         number;
}

const DEFAULT_POLL: PollConfig = {
  pollIntervalMs:        5000,
  flowingPollIntervalMs: 5000,
  idlePollIntervalMs:    30000,
  configVersion:         1,
};

export class Uplink {
  private buffer: RingBuffer<ReadingPayload>;
  private url:    string;

  constructor(private cfg: UplinkConfig) {
    this.buffer = new RingBuffer<ReadingPayload>(cfg.bufferCapacity ?? 60);
    this.url    = `${cfg.homehubUrl.replace(/\/$/, '')}/internal/sensors/flow/readings`;
  }

  /**
   * Enqueue + attempt to flush. Returns the latest poll config from the
   * server, or DEFAULT_POLL if the request didn't succeed.
   */
  async push(payload: ReadingPayload, now = Date.now()): Promise<PollConfig> {
    this.buffer.push(payload, now);
    return this.flush(now);
  }

  /** Drain buffer to server, oldest first. Stops on first failure (re-buffers). */
  async flush(now = Date.now()): Promise<PollConfig> {
    let lastConfig: PollConfig = DEFAULT_POLL;

    while (!this.buffer.isEmpty()) {
      const item = this.buffer.peek()!;
      const tOffsetMs = Math.max(0, now - item.enqueuedAt);

      try {
        lastConfig = await this.postOnce({
          ...item.value,
          // Append the tOffsetMs so the server can reconstruct recordedAt.
          // (Cast: tOffsetMs is a wire-only field, not part of ReadingPayload.)
        }, tOffsetMs);
        // Success — pop from the buffer and continue with next item.
        this.buffer.shift();
      } catch (err) {
        // Network failure — leave the item at the head of the buffer for next push.
        // Don't throw; the orchestrator should keep polling.
        if (typeof trace === 'function') {
          trace(`uplink: post failed, buffered (${this.buffer.length} pending): ${(err as Error).message}\n`);
        }
        return lastConfig;
      }
    }

    return lastConfig;
  }

  get pendingCount(): number {
    return this.buffer.length;
  }

  private async postOnce(payload: ReadingPayload, tOffsetMs: number): Promise<PollConfig> {
    const body = JSON.stringify({
      sensorId:         this.cfg.sensorId,
      rateM3h:          payload.rateM3h,
      totalM3:          payload.totalM3 ?? undefined,
      signalQuality:    payload.signalQuality ?? undefined,
      rssi:             payload.rssi ?? undefined,
      uptimeSec:        payload.uptimeSec ?? undefined,
      firmwareVersion:  payload.firmwareVersion ?? undefined,
      mac:              payload.mac ?? undefined,
      bootReason:       payload.bootReason ?? undefined,
      tOffsetMs,
    });

    const res = await fetch(this.url, {
      method: 'POST',
      headers: {
        'Content-Type':   'application/json',
        'X-Internal-Key': this.cfg.internalApiKey,
      },
      body,
    });

    if (!res.ok) {
      throw new Error(`HTTP ${res.status} ${res.statusText}`);
    }

    try {
      const json = await res.json() as Partial<PollConfig>;
      return {
        pollIntervalMs:        json.pollIntervalMs        ?? DEFAULT_POLL.pollIntervalMs,
        flowingPollIntervalMs: json.flowingPollIntervalMs ?? DEFAULT_POLL.flowingPollIntervalMs,
        idlePollIntervalMs:    json.idlePollIntervalMs    ?? DEFAULT_POLL.idlePollIntervalMs,
        configVersion:         json.configVersion         ?? DEFAULT_POLL.configVersion,
      };
    } catch {
      return DEFAULT_POLL;
    }
  }
}

File: src/firmware/c/main/uplink.c
/**
 * Pure helpers for uplink — JSON request body serializer, poll-config
 * default merging, ring buffer init. The HTTP path + cJSON response
 * decoding live in uplink_esp.c.
 */
#include <stdarg.h>
#include <stdio.h>
#include <string.h>

#include "uplink.h"

const uplink_poll_config_t UPLINK_DEFAULT_POLL_CONFIG = {
    .poll_interval_ms         = UPLINK_DEFAULT_POLL_INTERVAL_MS,
    .flowing_poll_interval_ms = UPLINK_DEFAULT_FLOWING_POLL_INTERVAL_MS,
    .idle_poll_interval_ms    = UPLINK_DEFAULT_IDLE_POLL_INTERVAL_MS,
    .config_version           = UPLINK_DEFAULT_CONFIG_VERSION,
};

/* Helper: append `chunk` to buf at *len. Returns false on overflow. */
static bool append(char *buf, size_t buf_size, int *len, const char *fmt, ...)
{
    int remaining = (int)buf_size - *len;
    if (remaining <= 0) return false;

    va_list args;
    va_start(args, fmt);
    int n = vsnprintf(buf + *len, (size_t)remaining, fmt, args);
    va_end(args);

    if (n < 0 || n >= remaining) return false;
    *len += n;
    return true;
}

int uplink_build_request_json(char *buf, size_t buf_size,
                              const uplink_config_t *cfg,
                              const uplink_reading_t *reading,
                              int64_t now_ms)
{
    if (buf == NULL || buf_size == 0 || cfg == NULL || reading == NULL) {
        return -1;
    }

    int64_t t_offset_ms = now_ms - reading->enqueued_at_ms;
    if (t_offset_ms < 0) t_offset_ms = 0;

    int len = 0;

    /* Required: sensor_id + rate_m3h + tOffsetMs */
    if (!append(buf, buf_size, &len,
                "{\"sensorId\":\"%s\",\"rateM3h\":%g,\"tOffsetMs\":%lld",
                cfg->sensor_id, (double)reading->rate_m3h, (long long)t_offset_ms)) {
        return -1;
    }

    if (reading->has_total_m3) {
        if (!append(buf, buf_size, &len, ",\"totalM3\":%g", (double)reading->total_m3))
            return -1;
    }
    if (reading->has_signal_quality) {
        if (!append(buf, buf_size, &len, ",\"signalQuality\":%d", reading->signal_quality))
            return -1;
    }
    if (reading->has_rssi) {
        if (!append(buf, buf_size, &len, ",\"rssi\":%d", reading->rssi))
            return -1;
    }
    if (reading->has_uptime_sec) {
        if (!append(buf, buf_size, &len, ",\"uptimeSec\":%lld",
                    (long long)reading->uptime_sec))
            return -1;
    }
    if (reading->firmware_version[0] != '\0') {
        if (!append(buf, buf_size, &len, ",\"firmwareVersion\":\"%s\"",
                    reading->firmware_version))
            return -1;
    }
    if (reading->mac[0] != '\0') {
        if (!append(buf, buf_size, &len, ",\"mac\":\"%s\"", reading->mac))
            return -1;
    }
    if (reading->boot_reason[0] != '\0') {
        if (!append(buf, buf_size, &len, ",\"bootReason\":\"%s\"", reading->boot_reason))
            return -1;
    }
    /* Always emitted — server uses it to decide whether to insert into
     * flow_readings or treat the POST as a diagnostic-only heartbeat. */
    if (!append(buf, buf_size, &len, ",\"meterReachable\":%s",
                reading->meter_reachable ? "true" : "false")) {
        return -1;
    }

    if (!append(buf, buf_size, &len, "}")) return -1;
    return len;
}

void uplink_apply_poll_defaults(uplink_poll_config_t *pc)
{
    if (pc == NULL) return;
    if (pc->poll_interval_ms         <= 0) pc->poll_interval_ms         = UPLINK_DEFAULT_POLL_INTERVAL_MS;
    if (pc->flowing_poll_interval_ms <= 0) pc->flowing_poll_interval_ms = UPLINK_DEFAULT_FLOWING_POLL_INTERVAL_MS;
    if (pc->idle_poll_interval_ms    <= 0) pc->idle_poll_interval_ms    = UPLINK_DEFAULT_IDLE_POLL_INTERVAL_MS;
    if (pc->config_version           <= 0) pc->config_version           = UPLINK_DEFAULT_CONFIG_VERSION;
}

void uplink_init(uplink_t *u, const uplink_config_t *cfg)
{
    if (u == NULL || cfg == NULL) return;
    memset(u, 0, sizeof(*u));
    u->cfg              = *cfg;
    u->last_poll_config = UPLINK_DEFAULT_POLL_CONFIG;
    ring_buffer_init(&u->buffer, u->buffer_storage,
                     sizeof(uplink_reading_t), UPLINK_BUFFER_CAPACITY);
}

size_t uplink_pending(const uplink_t *u)
{
    return (u == NULL) ? 0 : ring_buffer_length(&u->buffer);
}
